File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/modeljson/RelationshipPersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.modeljson;

import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmE2ERelationship;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.persistence.modeljson.types.AttributeReference;
import com.microsoft.commondatamodel.objectmodel.persistence.modeljson.types.SingleKeyRelationship;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.StringUtils;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * The relationship persistence.
 */
public class RelationshipPersistence {
  private static final String TAG = RelationshipPersistence.class.getSimpleName();
  public static CompletableFuture<CdmE2ERelationship> fromData(
      final CdmCorpusContext ctx,
      final SingleKeyRelationship obj,
      final Map<String, String> entityPathByName) {
    // The type attribute is not being set by the default serializer because of the Order property.
    // Since we just have one type for now this should be okay.
    /*if (obj.Type != "SingleKeyRelationship")
    {
      // We don't have any other type of a relationship yet!
      return null;
    }*/

    if (!entityPathByName.containsKey(obj.getFromAttribute().getEntityName())) {
      Logger.warning(ctx, TAG, "fromData", null, CdmLogCode.WarnPersistRelUndefinedSourceEntity , obj.getFromAttribute().getEntityName());
      return CompletableFuture.completedFuture(null);
    }
    if (!entityPathByName.containsKey(obj.getToAttribute().getEntityName())) {
      Logger.warning(ctx, TAG, "fromData", null, CdmLogCode.WarnPersistRelUndefinedSourceEntity , obj.getFromAttribute().getEntityName());
      return CompletableFuture.completedFuture(null);
    }

    final CdmE2ERelationship relationship = ctx.getCorpus()
        .makeObject(CdmObjectType.E2ERelationshipDef, obj.getName());

    return Utils.processAnnotationsFromData(ctx, obj, relationship.getExhibitsTraits())
        .thenApply(v -> {
          relationship.setExplanation(obj.getDescription());
          relationship
              .setFromEntity(entityPathByName.get(obj.getFromAttribute().getEntityName()));
          relationship.setToEntity(entityPathByName.get(obj.getToAttribute().getEntityName()));
          relationship.setFromEntityAttribute(obj.getFromAttribute().getAttributeName());
          relationship.setToEntityAttribute(obj.getToAttribute().getAttributeName());

          return relationship;
        });
  }

  public static CompletableFuture<SingleKeyRelationship> toData(
      final CdmE2ERelationship instance,
      final ResolveOptions resOpt,
      final CopyOptions options) {
    final AttributeReference fromAttribute = new AttributeReference();
    fromAttribute.setEntityName(getEntityName(instance.getFromEntity()));
    fromAttribute.setAttributeName(instance.getFromEntityAttribute());

    final AttributeReference toAttribute = new AttributeReference();
    toAttribute.setEntityName(getEntityName(instance.getToEntity()));
    toAttribute.setAttributeName(instance.getToEntityAttribute());

    final SingleKeyRelationship result = new SingleKeyRelationship();
    result.setType("SingleKeyRelationship");
    result.setDescription(instance.getExplanation());
    result.setName(instance.getName());
    result.setFromAttribute(fromAttribute);
    result.setToAttribute(toAttribute);

    Utils.processTraitsAndAnnotationsToData(instance.getCtx(), result, instance.getExhibitsTraits());
    
    return CompletableFuture.completedFuture(result);
  }

  private static String getEntityName(final String corpusPath) {
    final int lastSlashIndex = corpusPath.lastIndexOf("/");
    if (lastSlashIndex != -1) {
      return StringUtils.slice(corpusPath, lastSlashIndex + 1);
    }

    return corpusPath;
  }
}

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/ModelJson/RelationshipPersistence.cs
namespace Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.ModelJson.types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System.Collections.Generic;
    using System.Threading.Tasks;

    /// <summary>
    /// The relationship persistence.
    /// </summary>
    class RelationshipPersistence
    {
        public static async Task<CdmE2ERelationship> FromData(CdmCorpusContext ctx, SingleKeyRelationship obj, Dictionary<string, string> entitySchemaByName)
        {
            // TODO: re-add this once Dan's serializer is merged. 
            // The type attribute is not being set by the default serializer because of the Order property.
            // Since we just have one type for now this should be okay.
            /*if (obj.Type != "SingleKeyRelationship")
            {
                // We don't have any other type of a relationship yet!
                return null;
            }*/

            if (!entitySchemaByName.ContainsKey(obj.FromAttribute.EntityName))
            {
                Logger.Warning(nameof(RelationshipPersistence), ctx, $"Relationship's source entity '{obj.FromAttribute.EntityName}' is not defined.");

                return null;
            }

            if (!entitySchemaByName.ContainsKey(obj.ToAttribute.EntityName))
            {
                Logger.Warning(nameof(RelationshipPersistence), ctx, $"Relationship's target entity '{obj.ToAttribute.EntityName}' is not defined.");

                return null;
            }
            
            var relationship = ctx.Corpus.MakeObject<CdmE2ERelationship>(CdmObjectType.E2ERelationshipDef, obj.Name);

            await Utils.ProcessAnnotationsFromData(ctx, obj, relationship.ExhibitsTraits);

            relationship.Explanation = obj.Description;
            relationship.FromEntity = entitySchemaByName[obj.FromAttribute.EntityName];
            relationship.ToEntity = entitySchemaByName[obj.ToAttribute.EntityName];
            relationship.FromEntityAttribute = obj.FromAttribute.AttributeName;
            relationship.ToEntityAttribute = obj.ToAttribute.AttributeName;

            return relationship;
        }

        public static async Task<SingleKeyRelationship> ToData(CdmE2ERelationship instance, ResolveOptions resOpt, CopyOptions options)
        {
            var fromAttribute = new AttributeReference
            {
                EntityName = GetEntityName(instance.FromEntity),
                AttributeName = instance.FromEntityAttribute
            };

            var toAttribute = new AttributeReference
            {
                EntityName = GetEntityName(instance.ToEntity),
                AttributeName = instance.ToEntityAttribute
            };
            
            var result = new SingleKeyRelationship
            {
                Type = "SingleKeyRelationship",
                Description = instance.Explanation,
                Name = instance.Name,
                FromAttribute = fromAttribute,
                ToAttribute = toAttribute
            };

            await Utils.ProcessAnnotationsToData(instance.Ctx, result, instance.ExhibitsTraits);

            return result;
        }

        private static string GetEntityName(string corpusPath)
        {
            var lastSlashIndex = corpusPath.LastIndexOf("/");

            if (lastSlashIndex != -1)
            {
                return corpusPath.Slice(lastSlashIndex + 1);
            }

            return corpusPath;
        }
    }
}
